kit/Kit.cpp

```diff
@@ -3983,4 +4001,225 @@ void dump_kit_state()
+        //std::vector<lineOfUiLogFile> aLoadedLines;
+        while (bFileNotEnd || nProcessedLines<nLoadedLines)
+        {
+            // if the line to process is not even loaded yet
+            if (nLoadedLines == nProcessedLines)
+            {
+                if (!loadNextLine())
+                    break;
+            }
+            // We can process the actual line (nLoadedLines > nProcessedLines)
+            // If the actual line is not merged yet.
+            if (!aLoadedLines[nProcessedLines].bMerged)
+            {
+                // Some special cases needs to be converted:
+
+                // if undo count reached max, it will decrease 1 and then increase again
+                // remove the "false" decrease. If it is deleted, this line should not to be logged.
+                if (!removeIfFalseUndoDecrease(nProcessedLines))
+                {
+                    // merge mouse-down + mouse-up into mouse-click
+                    mergeIfMouseClick(nProcessedLines);
+
+                    //if the actual line is mergeable .. search all the other lines to merge into it.
+                    int nMergeCount = 1;
+                    int nActLine = nProcessedLines;
+                    bool bBreakSearch = false;
+                    int nLastMergedTime = 0;
+                    while(!bBreakSearch)
+                    {
+                        int nNextLine = findNextUserLine(nActLine);
+                        if (nNextLine >= 0)
+                        {
+                            if (aLoadedLines[nProcessedLines].aCmd == "cmd:mouse" && aLoadedLines[nProcessedLines].aSubCmd == "click")
+                            {
+                                mergeIfMouseClick(nNextLine);
+                            }
+
+                            // If the original line is a cmd, skip all undo change log, only new shoul break it.
+                            // If the actual line is deleted .. then search for next, beacuse
+                            // - if the original is a cmd, then the actual merged cmd must be a mouse-up (merged into click)
+                            // - if the original is an undo-count, then only a new non-merged cmd can break its merge.
+                            if (!(aLoadedLines[nNextLine].bMerged
+                                || (aLoadedLines[nProcessedLines].aCmd.starts_with("cmd")
+                                && aLoadedLines[nNextLine].aCmd.starts_with("undo-count-change"))))
+                            {
+                                if (aLoadedLines[nNextLine].aCmd == aLoadedLines[nProcessedLines].aCmd
+                                    && aLoadedLines[nNextLine].aSubCmd == aLoadedLines[nProcessedLines].aSubCmd)
+                                {
+                                    //found a mergeable .. continue to search more
+                                    nMergeCount++;
+                                    aLoadedLines[nNextLine].bMerged = true;
```

Comment:
I'm pretty unhappy with arithmetic on these aLoadedLines - which should be _loadedLines - but why do we load lines ?

Surely we can do event de-duplication and merging efficiently as we stream these to the log file.

Reply:
removed it as it was part of the post-process.
loading line is still present but now it is only a "std::getline(_fileStreamUICommands, line)"

loadedLines was a cache of lines preloaded from the temp file, before they saved into the permanent file.
I kept them in memory as merge needed some preloaded lines, but i made sure that every line loaded only once (and only when that was needed).. and kept memory usage at minimum possible for the merge.